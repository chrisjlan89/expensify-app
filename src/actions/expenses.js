import uuid from 'uuid';
import database from '../firebase/firebase';

//component calls action generator
//action generator return obj
//component dispatches obj
//redux store chnage


//component calls action generator
//action generator return function
//component dispatches function (?)
//function runs (has the ability to dispatch other actions and do w/e it wants )
 

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
       const {
        description = '', 
        note ='' , 
        amount = 0, 
        createdAt = 0
       } = expenseData;
       const expense = { description, note, amount, createdAt }
       
      return database.ref('expenses').push(expense).then((ref) =>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
       });
    };
}
//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type : 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({id}  = {}) => {
    console.log('this 1', id)
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(()=>{
             console.log(id, 'id')
             console.log( 'expense removed')
             dispatch(removeExpense({id}))
         })
     }
console.log('this')
}
//EDIT_EXPENse
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        })
    }
}

//SET_EXPENSES
export const setExpenses = (expenses) =>({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = ()=>  {
    return (dispatch) => {
    return database.ref('expenses').once('value').then((data) => {
       const expenses = [];
       data.forEach((childData)=> {
          expenses.push({
              id :childData.key,
              ...childData.val()
          });
       });
       console.log(expenses)
       dispatch(setExpenses(expenses));
    })
    }
  
} ;


