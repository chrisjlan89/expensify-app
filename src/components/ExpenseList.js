import React from 'react';
import {connect} from 'react-redux';
import filters from '../reducers/filters';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    
    <div>
       {
           props.expenses.length === 0 ? (
            <p> No expenses</p>
           ) : (
            props.expenses.map((expense, index) => (
                <ExpenseListItem key={expense.id} {...expense}/>
            ))
            

           )
       }
     
      {
                
                
            
            }
       
    </div>
)

const mapStateToProps = (state) => {
    console.log(state)
    return  {
       expenses : selectExpenses(state.expenses , state.filters)
    }
}

export default  connect(mapStateToProps)(ExpenseList);
