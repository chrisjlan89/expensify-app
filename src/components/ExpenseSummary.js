import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import filters from '../reducers/filters';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({expenseCount , expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'Expense' : 'Expenses'
    const expenseAmount = expensesTotal > 0 ?  numeral(expensesTotal / 100).format('$0,0.00') : ''
    return(
    <div>

       {
       
       expenseCount > 0 && <h3> Viewing {expenseCount} {expenseWord} totaling {expenseAmount} </h3>
    }

    {console.log(expenseCount)}
    {console.log(expensesTotal)}          
       
    </div>
)
}

const mapStateToProps = (state) => {
    console.log(state)
    const visibleExpenses = selectExpenses(state.expenses , state.filters)
    return  {
       expenseCount : visibleExpenses.length ,
       expensesTotal : expensesTotal(visibleExpenses)
    }
}

export default  connect(mapStateToProps)(ExpenseSummary);
