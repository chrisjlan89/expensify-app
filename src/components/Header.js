import React from 'react';
import {NavLink} from 'react-router-dom';



const Header = () => (
    <div>
     <header>
        <h1>Expensify</h1>
        <NavLink to ='/' activeClassName='is-active' exact ={true}> Dashboard Home </NavLink>
        <NavLink to ='/create'  activeClassName='is-active' exact ={true}> Add Expenses </NavLink>
        
        
     </header>
    </div>
);

export default Header;