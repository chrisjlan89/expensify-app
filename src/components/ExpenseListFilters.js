import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate , sortByAmount, setStartDate , setEndDate} from '../actions/filters'
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused : null,

    }

    onDatesChange = ({startDate , endDate}) => {
        this.props(setStartDate(startDate));
        this.props(setEndDate(endDate));
    }


    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) =>{
        props(setTextFilter(e.target.value));
    }

    onSortChange = (e) =>{
        let sValue = e.target.value; 
           sValue == 'date' ? this.props(sortByDate(sValue)) :
            this.props(sortByAmount(sValue))
        }

    render() {
        return(
            <div>
        <input type='text' 
        value={this.props.filters.text} 
        onChange={this.onTextChange} />
        <select 
        value={this.props.filters.sortBy}
        onChange={this.onSortChange} >
            <option value='date'>Date</option>
            <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
        startDate ={this.props.filters.startDate}
        endDate = {this.props.filters.endDate}
        onDatesChange = {this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
        />
    </div>
        )
    }
}

// check this file if sort isn't sorting properly

    

const mapStateToProps = (state) => ({
    filters : state.filters
    });


const mapDispatchToProps = (dispatch) => ({
    setTextFilter : (text) => dispatch(setTextFilter(text)),
    sortByDate : () => dispatch(sortByDate()),
    sortByAmount : () => dispatch(sortByAmount()),
    setStartDate : (startDate) => dispatch(setStartDate(startDate)),
    setEndDate : (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps)(ExpenseListFilters);