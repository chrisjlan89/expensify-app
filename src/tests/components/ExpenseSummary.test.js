import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary'
import expenses from '../fixtures/expenses';


test('should render  Expense Summary correctly  with 1 expense ', ()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={235}/>);
    expect(wrapper).toMatchSnapshot();
});


test('should render  Expense Summary correctly  with multiple expenses ', ()=>{
    const wrapper = shallow(<ExpenseSummary expenseCount={3} expensesTotal={2352302}/>);
    expect(wrapper).toMatchSnapshot();
})
