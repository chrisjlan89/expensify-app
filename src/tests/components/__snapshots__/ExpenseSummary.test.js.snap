// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render  Expense Summary correctly  with 1 expense  1`] = `
<div>
  <h3>
     Viewing 
    1
     
    Expense
     totaling 
    $2.35
     
  </h3>
</div>
`;

exports[`should render  Expense Summary correctly  with multiple expenses  1`] = `
<div>
  <h3>
     Viewing 
    3
     
    Expenses
     totaling 
    $23,523.02
     
  </h3>
</div>
`;
